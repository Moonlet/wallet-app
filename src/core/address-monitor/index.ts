import { Notifications } from '../messaging/notifications/notifications';
import { Blockchain } from '../blockchain/types';
import DeviceInfo from 'react-native-device-info';
import { IWalletsState, IAccountState } from '../../redux/wallets/state';
import { CONFIG } from '../../config';

const url = CONFIG.dataApiUrl + '/notifications/register';
const monitoredBlockchains = [Blockchain.ETHEREUM, Blockchain.ZILLIQA];

// const getDeviceToken = (): Promise<string> =>
//     Platform.OS === 'android' ? Notifications.getToken() : getApnsToken();

export const updateAddressMonitorTokens = (wallets: IWalletsState) => {
    const addresses = {};

    Object.keys(wallets || []).forEach(id => {
        wallets[id].accounts.forEach((account: IAccountState) => {
            if (monitoredBlockchains.indexOf(account.blockchain) !== -1) {
                if (!addresses[account.blockchain]) {
                    addresses[account.blockchain] = [];
                }

                addresses[account.blockchain].push(account.address);
            }
        });
    });

    const requestAddresses = Object.keys(addresses).map(blockchain => ({
        blockchain,
        addresses: addresses[blockchain]
    }));

    const deviceId = DeviceInfo.getUniqueId();
    Notifications.getToken().then(token => {
        const request = {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                deviceId,
                tokenType: 'fcm',
                token,
                addresses: requestAddresses
            })
        };

        fetch(url, request);
    });
};
