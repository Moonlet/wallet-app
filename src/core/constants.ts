import { Blockchain } from './blockchain/types';

const BLOCKCHAIN_COINS = {
    [Blockchain.ETHEREUM]: 'ETH',
    [Blockchain.ZILLIQA]: 'ZIL',
    [Blockchain.COSMOS]: 'ATOM',
    [Blockchain.STELLAR]: 'XLM'
};

export { BLOCKCHAIN_COINS };
