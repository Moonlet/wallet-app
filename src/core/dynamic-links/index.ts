import dynamicLinks from '@react-native-firebase/dynamic-links';
import { URL, URLSearchParams } from 'react-native-url-polyfill';
import { captureException as SentryCaptureException } from '@sentry/react-native';
import { IReduxState } from '../../redux/state';
import { NavigationService } from '../../navigation/navigation-service';
import { getChainId } from '../../redux/preferences/selectors';
import { setSelectedBlockchain } from '../../redux/wallets/actions';
import { Blockchain } from '../blockchain/types';
import { getNrPendingTransactions, getSelectedAccount } from '../../redux/wallets/selectors';
import { ITokenState } from '../../redux/wallets/state';
import { store } from '../../redux/config';

export class DynamicLinksService {
    private onLinkListener = null;
    private state: IReduxState = null;

    private handlers = {
        '/:blockchain/:token/validator/:validatorAddress': params => {
            this.navigateValidatorAddress(params);
        },
        '/:blockchain/:token/stake/:validatorAddress': params => {
            this.stakeValidatorAddress(params);
        },
        '/:blockchain/stake': params => {
            this.blockchainStake(params);
        },
        '/stake': params => {
            // TODO: implement later on
            // console.log('stakeDefault', params);
        }
    };

    private navigateValidatorAddress(params: any) {
        const blockchain = String(params?.blockchain)?.toUpperCase() as Blockchain;
        const validatorAddress = params?.validatorAddress;
        const tokenSymbol = String(params?.token)?.toUpperCase();

        if (blockchain && validatorAddress && tokenSymbol) {
            // make sure `blockchain` is selected
            store.dispatch(setSelectedBlockchain(blockchain));

            const chainId = getChainId(this.state, blockchain);

            const selectedAccount = getSelectedAccount(this.state);

            if (!selectedAccount) {
                SentryCaptureException(
                    new Error(
                        `Dynamic Links - no selected account, no wallet, blockchain: ${blockchain}, tokenSymbol: ${tokenSymbol}, validatorAddress: ${validatorAddress}`
                    )
                );
                return;
            }

            const token: ITokenState = selectedAccount.tokens[chainId][tokenSymbol];

            NavigationService.navigate('Validator', {
                blockchain,
                undefined, // validator
                accountIndex: selectedAccount.index,
                token,
                canPerformAction: !getNrPendingTransactions(this.state), // hasPendingTransactions
                options: {
                    validatorAddress,
                    tokenSymbol
                }
            });
        } else {
            SentryCaptureException(
                new Error(`Dynamic Links - navigateValidatorAddress, invalid params: ${params}`)
            );
        }
    }

    private stakeValidatorAddress(params: any) {
        const blockchain = String(params?.blockchain)?.toUpperCase() as Blockchain;
        const validatorAddress = params?.validatorAddress;

        if (blockchain && validatorAddress) {
            // make sure `blockchain` is selected
            store.dispatch(setSelectedBlockchain(blockchain));

            NavigationService.navigate('SmartScreen', {
                context: {
                    screen: 'StakeNow',
                    step: 'SelectStakeAccount',
                    key: 'select-stake-account',
                    params: {
                        validatorId: validatorAddress
                    }
                },
                navigationOptions: {
                    title: 'Stake now'
                },
                newFlow: true
            });
        } else {
            SentryCaptureException(
                new Error(`Dynamic Links - stakeValidatorAddress, invalid params: ${params}`)
            );
        }
    }

    private blockchainStake(params: any) {
        const blockchain = String(params?.blockchain)?.toUpperCase() as Blockchain;

        if (blockchain) {
            // make sure `blockchain` is selected
            store.dispatch(setSelectedBlockchain(blockchain));

            NavigationService.navigate('SmartScreen', {
                context: {
                    screen: 'StakeNow',
                    step: 'StakeSelectValidator',
                    key: 'stake-now-select-validator'
                },
                navigationOptions: {
                    title: 'Stake now'
                },
                newFlow: true
            });
        } else {
            SentryCaptureException(
                new Error(`Dynamic Links - blockchainStake, invalid params: ${params}`)
            );
        }
    }

    public async configure(state: IReduxState) {
        this.state = state;

        const dl = dynamicLinks();

        this.onLinkListener = dl.onLink(link => {
            this.handleDynamicLink(link);
        });

        // application is in a background state or has fully quit
        dl.getInitialLink().then(link => {
            this.handleDynamicLink(link);
        });
    }

    // app is in the foreground state
    public async handleDynamicLink(link: any) {
        if (link?.url) {
            this.processUrl(link.url, this.handlers, {
                protocol: ['https:'],
                host: [
                    'moonlet.io',
                    'moonlet.xyz',
                    'moonletwallet.com',
                    'moonlet.app',
                    'moonlet.link'
                ]
            });
        } else {
            // Invalid link url, link:${link}`
        }
    }

    public removeListeners() {
        typeof this.onLinkListener === 'function' && this.onLinkListener();
    }

    private pathPatternToRegexp(pathPattern: string) {
        let exp = '^';

        const pathParts = pathPattern.split('/').filter(Boolean);
        const capturingGroups = [];
        for (const part of pathParts) {
            exp += '\\/';
            if (part.indexOf(':') === 0) {
                exp += '([^\\/]*)';
                capturingGroups.push(part.replace(/^:/, ''));
            } else {
                exp += part;
            }
        }

        exp += '$';
        return {
            regexp: new RegExp(exp),
            capturingGroups
        };
    }

    private parsePath(path: string, pathPatternString: string) {
        const pathPattern = this.pathPatternToRegexp(pathPatternString);
        const match = path.match(pathPattern.regexp);

        if (match) {
            const params = {};
            for (let i = 0; i < pathPattern.capturingGroups.length; i++) {
                params[pathPattern.capturingGroups[i]] = match[i + 1];
            }
            return params;
        }

        return null;
    }

    private processPath(path: string, handlers: any, urlParams = {}) {
        const pathsPatterns = Object.keys(handlers);
        // console.log(path, handlers, urlParams);
        for (const pathPattern of pathsPatterns) {
            if (typeof handlers[pathPattern] === 'function') {
                const params = this.parsePath(path, pathPattern);
                if (params) {
                    return handlers[pathPattern]({
                        ...urlParams,
                        ...params
                    });
                }
            }
        }
    }

    private paramsToObject(entries: any) {
        const result = {};
        for (const [key, value] of entries) {
            // each 'entry' is a [key, value] tupple
            result[key] = value;
        }
        return result;
    }

    private processUrl(url: string, pathHandlers: any = this.handlers, whitelist = {}) {
        const u = new URL(url);

        for (const key of Object.keys(whitelist)) {
            if (u && typeof u[key] === 'string' && whitelist[key].indexOf(u[key]) < 0) {
                SentryCaptureException(
                    new Error(
                        JSON.stringify(
                            `whitelist not matched ${key}=${u[key]} not in whitelist: ${
                                typeof whitelist[key] === 'string'
                                    ? whitelist[key]
                                    : whitelist[key].join(', ')
                            }`
                        )
                    )
                );
                return null;
            }
        }

        let path = u?.pathname;

        if (path && String(path).startsWith('/links')) path = path.replace('/links', '');

        const searchParams = new URLSearchParams(url);

        return this.processPath(path, pathHandlers, this.paramsToObject(searchParams.entries()));
    }
}

export const DynamicLinks = new DynamicLinksService();
