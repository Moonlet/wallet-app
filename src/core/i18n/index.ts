export * from './translation/translate-component';
export * from './translation/translate';
