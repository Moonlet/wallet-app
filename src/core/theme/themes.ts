export enum Theme {
    dark = 'dark',
    light = 'light'
}
