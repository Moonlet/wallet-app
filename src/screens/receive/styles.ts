import { StyleSheet } from 'react-native';
import { ITheme } from '../../core/theme/itheme';
import { BASE_DIMENSION } from '../../styles/dimensions';

export default (theme: ITheme) =>
    StyleSheet.create({
        container: {
            flex: 1,
            padding: BASE_DIMENSION * 2,
            paddingTop: BASE_DIMENSION * 4,
            flexDirection: 'column',
            backgroundColor: theme.colors.appBackground
        },
        qrCodeContainer: {
            flex: 1,
            flexDirection: 'column',
            justifyContent: 'center',
            alignItems: 'center',
            marginBottom: BASE_DIMENSION * 4
        },
        bottomButton: {
            marginHorizontal: BASE_DIMENSION * 2,
            marginBottom: BASE_DIMENSION * 3
        }
    });
