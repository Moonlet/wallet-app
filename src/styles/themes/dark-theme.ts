import { ITheme } from '../../core/theme/itheme';
import { COLORS } from '../colors';

export const darkTheme: ITheme = {
    dark: true,
    shadowGradient: ['#00000000', '#000000CC'],

    fontSize: {
        small: 12,
        regular: 16,
        large: 34
    },

    colors: {
        primary: COLORS.GRAY,
        accent: COLORS.AQUA,
        text: COLORS.WHITE,
        textDarker: COLORS.LIGHT_GRAY,
        positive: COLORS.HOT_GREEN,
        negative: COLORS.HOT_RED,

        cardBackground: COLORS.DARK_GRAY,
        appBackground: COLORS.COD_GRAY,
        settingsDivider: COLORS.BLACK
    }
};
