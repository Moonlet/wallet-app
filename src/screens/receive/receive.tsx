import React, { useState } from 'react';
import { View, Clipboard, Dimensions } from 'react-native';
import stylesProvider from './styles';
import { withTheme, IThemeProps } from '../../core/theme/with-theme';
import { Button } from '../../library/button/button';
import { smartConnect } from '../../core/utils/smart-connect';
import { translate } from '../../core/i18n';
import QRCode from 'react-native-qrcode-svg';
import { withNavigationParams, INavigationProps } from '../../navigation/with-navigation-params';
import { IAccountState } from '../../redux/wallets/state';
import { AccountAddress } from '../../components/account-address/account-address';
import { IReduxState } from '../../redux/state';
import { connect } from 'react-redux';
import { Blockchain } from '../../core/blockchain/types';
import { getAccount } from '../../redux/wallets/selectors';
import { HeaderLeftClose } from '../../components/header-left-close/header-left-close';
import { ITokenConfig } from '../../core/blockchain/types/token';
import { BASE_DIMENSION } from '../../styles/dimensions';

export interface INavigationParams {
    accountIndex: number;
    blockchain: Blockchain;
    token: ITokenConfig;
}

export interface IReduxProps {
    account: IAccountState;
}

export const mapStateToProps = (state: IReduxState, ownProps: INavigationParams) => {
    return {
        account: getAccount(state, ownProps.accountIndex, ownProps.blockchain)
    };
};

const navigationOptions = ({ navigation }: any) => ({
    headerLeft: <HeaderLeftClose navigation={navigation} />,
    title: translate('App.labels.receive')
});

export const ReceiveScreenComponent = (
    props: INavigationProps<INavigationParams> &
        IReduxProps &
        IThemeProps<ReturnType<typeof stylesProvider>>
) => {
    const [copied, setCopied] = useState(false);

    return (
        <View style={props.styles.container}>
            <AccountAddress account={props.account} token={props.navigation.state.params.token} />
            <View style={props.styles.qrCodeContainer}>
                <QRCode
                    value={props.account.address}
                    size={Dimensions.get('window').width - BASE_DIMENSION * 10}
                />
            </View>

            <Button
                testID="copy-clipboard"
                style={props.styles.bottomButton}
                onPress={() => {
                    Clipboard.setString(props.account.address);
                    setCopied(true);
                }}
            >
                {copied
                    ? translate('App.buttons.copiedBtn')
                    : translate('App.buttons.clipboardBtn')}
            </Button>
        </View>
    );
};

ReceiveScreenComponent.navigationOptions = navigationOptions;

export const ReceiveScreen = smartConnect(ReceiveScreenComponent, [
    connect(mapStateToProps, null),
    withTheme(stylesProvider),
    withNavigationParams()
]);
