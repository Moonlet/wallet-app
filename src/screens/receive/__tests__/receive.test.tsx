import React from 'react';
import { shallow } from 'enzyme';
import { ReceiveScreenComponent, IProps } from '../receive';
import stylesProvider from '../styles';
import { darkTheme } from '../../../styles/themes/dark-theme';
import { loadTranslations } from '../../../core/i18n';

const props: IProps = {
    // @ts-ignore
    navigation: {
        navigate: jest.fn(),
        state: {
            params: {
                token: {}
            }
        }
    },
    styles: stylesProvider(darkTheme),
    theme: darkTheme,
    account: { address: 'value' }
};

export default describe('SendScreen', () => {
    beforeAll(async () => {
        await loadTranslations('en');
    });

    test('renders correctly', () => {
        const wrapper = shallow(<ReceiveScreenComponent {...props} />);
        expect(wrapper.debug()).toMatchSnapshot();
    });

    test('renders correctly - button should show Copied in title', () => {
        const wrapper = shallow(<ReceiveScreenComponent {...props} />);

        wrapper.find('[testID="copy-clipboard"]').simulate('Press');
        expect(wrapper.debug()).toMatchSnapshot();
    });
});
