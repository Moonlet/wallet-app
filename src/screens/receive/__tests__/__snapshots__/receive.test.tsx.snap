// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendScreen renders correctly - button should show Copied in title 1`] = `
"<View style={{...}}>
  <withTheme(AccountAddressComponent) account={{...}} token={{...}} />
  <View style={{...}}>
    <QRCode value=\\"value\\" size={670} color=\\"black\\" backgroundColor=\\"white\\" logoSize={20} logoBackgroundColor=\\"white\\" logoMargin={2} logoBorderRadius={0} ecl=\\"M\\" onError={[undefined]} gradientDirection={{...}} linearGradient={{...}} enableLinearGradient={false} quietZone={0} />
  </View>
  <withTheme(ButtonComponent) testID=\\"copy-clipboard\\" style={{...}} onPress={[Function: onPress]}>
    Copied
  </withTheme(ButtonComponent)>
</View>"
`;

exports[`SendScreen renders correctly 1`] = `
"<View style={{...}}>
  <withTheme(AccountAddressComponent) account={{...}} token={{...}} />
  <View style={{...}}>
    <QRCode value=\\"value\\" size={670} color=\\"black\\" backgroundColor=\\"white\\" logoSize={20} logoBackgroundColor=\\"white\\" logoMargin={2} logoBorderRadius={0} ecl=\\"M\\" onError={[undefined]} gradientDirection={{...}} linearGradient={{...}} enableLinearGradient={false} quietZone={0} />
  </View>
  <withTheme(ButtonComponent) testID=\\"copy-clipboard\\" style={{...}} onPress={[Function: onPress]}>
    Copy to clipboard
  </withTheme(ButtonComponent)>
</View>"
`;
